### PR TITLE
Mj fix amp epic problems

### PR DESCRIPTION
**Pull request by michaelbjacobson**
## What does this change?
- Prevents specific variable tokens (eg. `%%ARTICLE_COUNT%%`) being used on unsupported platforms.
- Web preview button disabled for all unsupported platforms (AMP, Apple News, Liveblog)

### CODE REVIEW

File: public/src/components/channelManagement/helpers/shared.ts [hunk 1]
```diff
@@ -5,6 +5,8 @@ export interface Variant {
   name: string;
 }
 
+export type TestPlatform = 'AMP' | 'DOTCOM';
```
Comment (tomrf1) [inline public/src/components/channelManagement/helpers/shared.ts:8]: The apple news epic tool also needs validation - it only supports `CURRENCY_SYMBOL` - https://github.com/guardian/apple-news/blob/73bf7b48d02b721425e9b80b3c17a222f8d0a288/app/integration/EpicService.scala

File: public/src/components/channelManagement/helpers/shared.ts [hunk 2]
```diff
@@ -28,6 +30,7 @@ export interface EpicEditorConfig {
   allowViewFrequencySettings: boolean;
   allowArticleCount: boolean;
   testNamePrefix?: string;
+  platform?: TestPlatform;
```
[the remark this answers is not in the document]
Comment (TomPretty) [reply public/src/components/channelManagement/helpers/shared.ts:33]: Yeah agreed!

[the remark this answers is not in the document]
Comment (michaelbjacobson) [reply public/src/components/channelManagement/helpers/shared.ts:33]: ✅

File: public/src/components/channelManagement/testEditorVariantSummaryWebPreviewButton.tsx [hunk 3]
```diff
@@ -1,39 +1,55 @@
...
 
-const PROD_BASE_ARTICLE_URL =
-  'https://theguardian.com/world/2020/may/08/commemorating-ve-day-during-coronavirus-lockdown-somehow-the-quiet-made-it-louder';
-
-const CODE_BASE_ARTICLE_URL =
-  'https://m.code.dev-theguardian.com/world/2020/may/08/commemorating-ve-day-during-coronavirus-lockdown-somehow-the-quiet-made-it-louder';
+const BASE_ARTICLE_URL = {
+  PROD: {
+    AMP:
+      'https://amp.theguardian.com/world/2020/may/08/commemorating-ve-day-during-coronavirus-lockdown-somehow-the-quiet-made-it-louder',
```
[the remark this answers is not in the document]
Comment (michaelbjacobson) [reply public/src/components/channelManagement/testEditorVariantSummaryWebPreviewButton.tsx]: Screenshot 2021-10-13 at 12 28 52

[the remark this answers is not in the document]
Comment (michaelbjacobson) [reply public/src/components/channelManagement/testEditorVariantSummaryWebPreviewButton.tsx]: It feels nicer to explicitly say it's not available for the given platform, as opposed to just not showing it. Users might assume the latter is a bug.

File: public/src/components/channelManagement/epicTests/epicTestVariantEditor.tsx [hunk 4]
```diff
@@ -72,6 +76,11 @@ const EpicTestVariantEditor: React.FC<EpicTestVariantEditorProps> = ({
 }: EpicTestVariantEditorProps) => {
   const classes = getUseStyles(epicEditorConfig.allowMultipleVariants)();
 
+  const templateValidator =
+    epicEditorConfig.platform && epicEditorConfig.platform === 'AMP'
+      ? ampInvalidTemplateValidator
+      : invalidTemplateValidator;
```
Comment (TomPretty) [inline public/src/components/channelManagement/epicTests/epicTestVariantEditor.tsx:82]: We can drop the initial `epicEditorConfig.platform &&` check now it's not optional, but do we need to check for `epicEditorConfig.platform === 'APPLE_NEWS'` here?

Comment (michaelbjacobson) [reply public/src/components/channelManagement/epicTests/epicTestVariantEditor.tsx]: ✅

File: public/src/components/channelManagement/helpers/validation.tsx [hunk 5]
```diff
@@ -93,9 +94,16 @@ export const CURRENCY_TEMPLATE = '%%CURRENCY_SYMBOL%%';
...
+const ampValidTemplates = [CURRENCY_TEMPLATE, COUNTRY_NAME_TEMPLATE];
+const appleNewsValidTemplates = [CURRENCY_TEMPLATE];
 
-export const invalidTemplateValidator = (text: string): string | boolean => {
+type TemplateValidatorForPlatform = (text: string) => string | boolean;
+
+const genericInvalidTemplateValidator = (
+  text: string,
+  validTemplates: string[],
+): string | boolean => {
```
[the remark this answers is not in the document]
Comment (michaelbjacobson) [reply public/src/components/channelManagement/helpers/validation.tsx:106]: Nice! I like this a lot!

[the remark this answers is not in the document]
Comment (michaelbjacobson) [reply public/src/components/channelManagement/helpers/validation.tsx]: I've implemented this now @TomPretty, all credit to you for the neat refactor! ✅

File: public/src/components/channelManagement/helpers/shared.ts [hunk 6]
```diff
@@ -5,6 +5,8 @@ export interface Variant {
   name: string;
 }
 
+export type TestPlatform = 'AMP' | 'APPLE_NEWS' | 'ARTICLE' | 'LIVEBLOG';
```
[the remark this answers is not in the document]
Comment (tomrf1) [reply public/src/components/channelManagement/helpers/shared.ts:8]: yep it's great that you've done this!

**Review by tomrf1 — APPROVED** [review]
great work!

**Comment by michaelbjacobson** [comment]
@tomrf1 @TomPretty any suggestions about the failing build? Running `build` locally works fine, but does throw up some webpack warnings about performance.

Screenshot 2021-10-15 at 13 24 37

File: public/src/components/channelManagement/helpers/shared.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (TomPretty) [reply public/src/components/channelManagement/helpers/shared.ts]: Yeah makes sense!

File: public/src/components/channelManagement/testEditorVariantSummaryWebPreviewButton.tsx [hunk 7]
```diff
@@ -42,12 +61,14 @@ const TestEditorVariantSummaryWebPreviewButton: React.FC<TestEditorVariantSummar
...
       onFocus={(event): void => event.stopPropagation()}
-      href={getPreviewUrl(testName, name, testType)}
+      href={getPreviewUrl(testName, name, testType, platform)}
       target="_blank"
       rel="noopener noreferrer"
-      disabled={isDisabled}
+      disabled={isCompatiblePlatform(platform) ? isDisabled : true}
     >
-      Web preview
+      {isCompatiblePlatform
```
[the remark this answers is not in the document]
Comment (michaelbjacobson) [reply public/src/components/channelManagement/testEditorVariantSummaryWebPreviewButton.tsx:69]: Aha! That was IntelliJ trying to be clever!

**Pull request merged**